File: packages/typescript/src/transport/rpc.ts
// packages/typescript/src/transport/rpc.ts
import { AEPError, AEPTransportError, fromJsonRpcError } from "../errors.js";
import { HEADER_PROTOCOL_VERSION, HEADER_SESSION_ID, PROTOCOL_VERSION, RPC_PATH } from "../constants.js";

export interface AuthBearer {
  bearer: string;
}

export interface TransportConfig {
  baseUrl: string;
  auth: AuthBearer;
  fetch?: typeof fetch;
}

interface JsonRpcRequest {
  jsonrpc: "2.0";
  id: number;
  method: string;
  params?: unknown;
}

interface JsonRpcResponse<T> {
  jsonrpc: "2.0";
  id: number;
  result?: T;
  error?: { code: number; message: string; data?: unknown };
}

export class RpcClient {
  private nextId = 1;
  private sessionId: string | null = null;
  private readonly baseUrl: string;
  private readonly bearer: string;
  private readonly fetchImpl: typeof fetch;

  constructor(config: TransportConfig) {
    this.baseUrl = config.baseUrl.replace(/\/$/, "");
    this.bearer = config.auth.bearer;
    this.fetchImpl = config.fetch ?? fetch;
  }

  setSessionId(sessionId: string): void {
    this.sessionId = sessionId;
  }

  getSessionId(): string | null {
    return this.sessionId;
  }

  async call<T>(method: string, params?: unknown): Promise<T> {
    const id = this.nextId++;
    const body: JsonRpcRequest = { jsonrpc: "2.0", id, method };
    if (params !== undefined) body.params = params;

    const headers: Record<string, string> = {
      "Content-Type": "application/json",
      "Accept": "application/json",
      "Authorization": `Bearer ${this.bearer}`,
      [HEADER_PROTOCOL_VERSION]: PROTOCOL_VERSION,
    };
    if (this.sessionId) headers[HEADER_SESSION_ID] = this.sessionId;

    let resp: Response;
    try {
      resp = await this.fetchImpl(`${this.baseUrl}${RPC_PATH}`, {
        method: "POST",
        headers,
        body: JSON.stringify(body),
      });
    } catch (e) {
      throw new AEPTransportError("Network error during RPC call", e);
    }

    if (!resp.ok && resp.status >= 500) {
      throw new AEPTransportError(`HTTP ${resp.status} ${resp.statusText}`);
    }

    let payload: JsonRpcResponse<T>;
    try {
      payload = (await resp.json()) as JsonRpcResponse<T>;
    } catch (e) {
      throw new AEPTransportError("Failed to parse JSON-RPC response", e);
    }

    if (payload.error) {
      throw fromJsonRpcError(payload.error);
    }
    if (payload.result === undefined) {
      throw new AEPTransportError("JSON-RPC response missing result and error");
    }
    return payload.result;
  }
}

File: packages/python/helix_sdk/transport/rpc.py
# packages/python/helix_sdk/transport/rpc.py
from __future__ import annotations

import json
from typing import Any

import httpx

from ..constants import HEADER_PROTOCOL_VERSION, HEADER_SESSION_ID, PROTOCOL_VERSION, RPC_PATH
from ..errors import AEPTransportError, from_jsonrpc_error


class RpcClient:
    def __init__(self, base_url: str, bearer: str, http_client: httpx.AsyncClient | None = None) -> None:
        self._base_url = base_url.rstrip("/")
        self._bearer = bearer
        self._next_id = 1
        self._session_id: str | None = None
        self._client = http_client or httpx.AsyncClient(timeout=httpx.Timeout(60.0, read=None))
        self._owns_client = http_client is None

    async def aclose(self) -> None:
        if self._owns_client:
            await self._client.aclose()

    def set_session_id(self, session_id: str) -> None:
        self._session_id = session_id

    def get_session_id(self) -> str | None:
        return self._session_id

    async def call(self, method: str, params: Any | None = None) -> Any:
        req_id = self._next_id
        self._next_id += 1
        body: dict[str, Any] = {"jsonrpc": "2.0", "id": req_id, "method": method}
        if params is not None:
            body["params"] = params

        headers = {
            "Content-Type": "application/json",
            "Accept": "application/json",
            "Authorization": f"Bearer {self._bearer}",
            HEADER_PROTOCOL_VERSION: PROTOCOL_VERSION,
        }
        if self._session_id:
            headers[HEADER_SESSION_ID] = self._session_id

        try:
            resp = await self._client.post(
                f"{self._base_url}{RPC_PATH}",
                headers=headers,
                content=json.dumps(body),
            )
        except httpx.HTTPError as e:
            raise AEPTransportError("Network error during RPC call", cause=e) from e

        if resp.status_code >= 500:
            raise AEPTransportError(f"HTTP {resp.status_code} {resp.reason_phrase}")

        try:
            payload = resp.json()
        except json.JSONDecodeError as e:
            raise AEPTransportError("Failed to parse JSON-RPC response", cause=e) from e

        if "error" in payload and payload["error"]:
            raise from_jsonrpc_error(payload["error"])
        if "result" not in payload:
            raise AEPTransportError("JSON-RPC response missing result and error")
        return payload["result"]
